File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem44_NumberOfProvinces.java
package org.phoenix.leetcode.challenges;

public class Problem44_NumberOfProvinces {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        int components = 0;
        boolean[] visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                components++;
                dfs(i, isConnected, visited);
            }
        }
        return components;
    }

    public void dfs(int currNode, int[][] isConnected, boolean[] visit) {
        visit[currNode] = true;
        for (int i = 0; i < isConnected.length; i++) {
            if (isConnected[currNode][i] == 1 && !visit[i]) {
                dfs(i, isConnected, visit);
            }
        }
    }
}

File: JavaScript/leetcode75/src/main/Problem44_NumberOfProvinces.js
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
var findCircleNum = function (isConnected) {
  const n = isConnected.length;
  let components = 0;
  let visited = new Array(n);
  for (let i = 0; i < n; i++) {
    if (!visited[i]) {
      components++;
      dfs(i, isConnected, visited);
    }
  }
  return components;
};

let dfs = function (currNode, isConnected, visited) {
  visited[currNode] = true;
  for (let i = 0; i < isConnected.length; i++) {
    if (isConnected[currNode][i] === 1 && !visited[i]) {
      dfs(i, isConnected, visited);
    }
  }
};

module.exports = findCircleNum;
